azure-functions-powershell-library/BindingExtractor.cs

```diff
@@ -0,0 +1,124 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using AzureFunctions.PowerShell.SDK.BundledBindings;
+using Common;
+using System.Management.Automation.Language;
+using System.Reflection;
+
+namespace AzureFunctions.PowerShell.SDK
+{
+    internal static class BindingExtractor
+    {
+        static BindingExtractor()
+        {
+            // We want this behavior to be extensible by third parties who may want to add binding definitions. 
+            // To accomplish this, we parse through all loaded assemblies to find any class which implements 
+            //   IBinding, and load it manually into our supportedBindings
+
+            // Find all types that implement IBinding
+            // Apparently, there are scenarios where this throws a ReflectionTypeLoadException - filtering it seems to help
+            IEnumerable<Assembly>? assemblies = AppDomain.CurrentDomain.GetAssemblies();
+            List<Type> types = new List<Type>();
+            foreach (Assembly assembly in assemblies)
+            {
+                try
+                {
+                    types.AddRange(assembly.GetTypes().Where(type => type.IsSubclassOf(typeof(IBinding))));
+                }
+                catch (ReflectionTypeLoadException)
+                {
+                    //Do nothing, this is probably fine (?)
+                }
+            }
+            foreach (Type type in types)
+            {
+                try
+                {
+                    // Instantiate an object of this type, and call its method to load it into supportedBindings for use later
+                    object? obj = Activator.CreateInstance(type);
+                    if (obj is not null)
+                    {
+                        ((IBinding)obj).AddToExtractor();
+                    }
+                }
+                catch (MissingMethodException)
+                {
+                    //Do nothing, it's an abstract class or improperly declared. 
+                }
+            }
+        }
+        private static List<IBinding> supportedBindings = new List<IBinding>();
```

Comment:
Wouldn't HashSet still allow duplicates? Unless it's a HashSet of Type, new instances of a type are treated as unique, right? Anyway, addSupportedBinding and hasSupportedBinding are actually already overkill, since the only time this collection is modified is when the static constructor runs, which in .NET is thread-safe and only runs once.